enhancements/update/eus-upgrades-mco.md

```diff
@@ -0,0 +1,171 @@
+---
+title: eus-upgrades-mco
+authors:
+  - "@rphillips"
+reviewers:
+  - @kikisdeliveryservice
+  - @yuqi-zhang
+approvers:
+  - "@derekwaynecarr"
+  - "@crawford"
+creation-date: 2021-05-05
+last-updated: 2021-05-05
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# EUS Upgrades MCO
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+- [x] [Enhancement 762](https://github.com/openshift/enhancements/pull/762)
+
+
+## Summary
+
+This enhancement outlines the Machine Config Operator (MCO) behavior when
+propogating Kubernetes Events in EUS upgrade scenarios.  The MCO will emit
+events that are user readable and be clear enough that a user can take an
+action.
+
+Related to [Enhancement 762](https://github.com/openshift/enhancements/pull/762).
+
+## Motivation
+
+The introduction of EUS creates a subset of clusters which we expect will run
+4.6 for a year or more then upgrade rapidly, though serially, from 4.6 to 4.10.
+This rapid upgrade introduces the risk that those clusters may upgrade faster
+than is safe due to constraints imposed by OpenShift, the upstream components of
+OpenShift, or deployed workloads.
+
+This also creates a scenario where admins wish to reduce both the duration and
+the disruption to workload associated with the upgrade.
+
+### Goals
+
+- The MCO will look at all the pools within the cluster, if there are paused
+pools the MCO will make sure the nodes have a n-2 skew.
+- If there is a paused pool with a node skew of greater than n-2, then an
+event will be emitted.
```

Comment:
Will this block the control plane (master) pool from upgrading or not?

An event seems odd.  If the goal is to block, then I think we need to change the MCO to go `Upgradable=False` so the CVO will refuse upgrades, right?
If the goal is to *not* block...well, some sort of important/critical alert seems in order, not just an event right?

Reply:
> If the goal is to block...

The API-server is in charge of blocking; see #762.  The MCO is supposed to remind folks that they have old nodes, and provide suggestions about getting them updated.  I'd expected that to be via alerts, but I guess spitting out events periodically could accomplish the same thing.

Reply:
Looks like implementation is moving forward in https://github.com/openshift/machine-config-operator/pull/2658
I guess the assumption is that we'll follow up with alerts based on the events?

Reply:
I think current implementation PR is more generic, it will send event if any node have skew >= 2